programs/vote/src/vote_state/mod.rs

```diff
@@ -1161,10 +1162,22 @@ pub fn withdraw<S: std::hash::BuildHasher>(
         .lamports()?
         .checked_sub(lamports)
         .ok_or(InstructionError::InsufficientFunds)?;
+    let received_credits_previous_epoch = match vote_state.epoch_credits.len() {
+        0..=1 => false,
+        _ => {
+            let (_epoch, epoch_credits, previous_epoch_credits) =
+                vote_state.epoch_credits[vote_state.epoch_credits.len() - 2];
+            epoch_credits.saturating_sub(previous_epoch_credits) > 0
+        }
+    };
```

Comment:
I'm sorry, I totally led you astray. I've been looking more carefully at how credits are incremented (thanks to your excellent test cases, and `with(out)_zero_credit_epoch` in particular). I thought that the epoch_credits vec was extended on each epoch boundary, but it looks like `VoteState::increment_credits()` is the only method that pushes to it: https://github.com/solana-labs/solana/blob/aea8f0df1610248d29d8ca3bc0d60e9fabc99e31/programs/vote/src/vote_state/mod.rs#L837-L862
This means that a validator that doesn't vote for an entire epoch won't have an epoch_credits entry for that epoch at all.

(Incidentally, I'm now confused about how we would ever hit this case, since every time a new entry is pushed, the observed credits are incremented: https://github.com/solana-labs/solana/blob/aea8f0df1610248d29d8ca3bc0d60e9fabc99e31/programs/vote/src/vote_state/mod.rs#L850
Sadly, no hint in the originating PR: https://github.com/solana-labs/solana/pull/7909)

Given that, I think that what we actually want to check is that the epoch in vote_state.epoch_credits.last() is <= current epoch - 2
Does that make sense?
We'll need to grab the Clock sysvar to read the current epoch.

Reply:
Thanks for the catch!

I've refactored to use the clock sysvar and check the current epoch. There are two places I'm unsure about (I've left comments here in the PR). I'm going to try debugging a running validator tomorrow to confirm some assumptions but any feedback is appreciated!